tests/sterm/test_terminal.py

```diff
@@ -0,0 +1,213 @@
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().join.assert_any_call(True)
+
+
+def test_joins_tx_thread_after_keyboard_interrupt(mock_sterm, mock_serial):
+    mock_sterm().join.side_effect = (KeyboardInterrupt(), None)
+
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().join.assert_called_with()
+
+
+def test_closes_sterm(mock_sterm, mock_serial):
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().close.assert_called_once()
+
+
+def test_closes_sterm_after_exception(mock_sterm, mock_serial):
+    mock_sterm().join.side_effect = (Exception(), None)
+    with pytest.raises(Exception):
+        terminal.run("tty.122", 9600)
+
+    mock_sterm().close.assert_called_once()
+
+
+def test_sets_terminal_special_chars(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+
+    assert term.exit_character == terminal.CTRL_C
+    assert term.menu_character == terminal.CTRL_T
+    assert term.reset_character == terminal.CTRL_B
+    assert term.help_character == terminal.CTRL_H
+
+
+def test_sets_terminal_rx_and_tx_encoding_to_utf8(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+
+    assert term.input_encoding == "UTF-8"
+    assert term.output_encoding == "UTF-8"
+
+
+def test_stops_terminal_when_ctrl_c_received(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.alive = True
+    mock_console().getkey.return_value = terminal.CTRL_C
+
+    term.writer()
+
+    assert term.alive is False
+
+
+def test_stops_terminal_on_keyboard_interrupt(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.alive = True
+    mock_console().getkey.side_effect = KeyboardInterrupt()
+
+    term.writer()
+
+    assert term.alive is False
+
+
+@pytest.mark.parametrize("menu_key", terminal.VALID_MENU_KEYS)
+def test_handles_valid_menu_key(menu_key, mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.handle_menu_key = mock.Mock()
+    term.alive = True
+    mock_console().getkey.side_effect = (terminal.CTRL_T, menu_key, terminal.CTRL_C)
+
+    term.writer()
+
+    term.handle_menu_key.assert_called_once_with(menu_key)
+
+
+INVALID_MENU_KEYS = tuple(set(chr(i) for i in range(0, 127)) - set(terminal.VALID_MENU_KEYS) - set([terminal.CTRL_H]))
+
+
+@pytest.mark.parametrize("menu_key", INVALID_MENU_KEYS)
+def test_ignores_invalid_menu_key(menu_key, mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.handle_menu_key = mock.Mock()
+    term.alive = True
+    mock_console().getkey.side_effect = (terminal.CTRL_T, menu_key)
+
+    with pytest.raises(StopIteration):
+        term.writer()
+
+    term.handle_menu_key.assert_not_called()
+
+
+def test_reset_sends_serial_break(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+
+    term.reset()
+
+    mock_serial().sendBreak.assert_called_once()
```

Comment:
Do we really need to send break conditions? Is this a reliable way to reset a target? @jamesbeyond Do you know if we have any targets that reset when given a serial break? Is this a common thing?

Reply:
Not sure, this was just what the old tool did. Serial break does reset my DISCO_L475 though.

Reply:
Note: If we didn't send serial break and change the menu text and key handling, we wouldn't need to subclass `Miniterm` at all. However, we will change the UX if we do that.

Reply:
Alright. Learn a new feature everyday. If we run --sterm without --flash, I can see it being useful that the target would be reset. I would have expected a separate reset mechanism than a serial break condition, but it works how it works.